### PR TITLE
AOS-38: Service Pair Naming Style

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

### CODE REVIEW

File: rfcs/38-service-pair-naming-style.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+updated_at: 2021-04-21
+---
+
+# AOS-38: Service Pair Naming Style
+
+## Background
+
+There are 10 services have been implemented and more services are on the way, and every service may have their only pairs. We need to design a naming style that meets following needs:
+
+- No ambiguity: the pair should be confused with other pairs.
```
Comment (Prnyself) [inline rfcs/38-service-pair-naming-style.md:13]: should not be confused?

**Comment by xxchan** [comment]
Why this rfc is numbered 38?

**Comment by xxchan** [comment]
BTW, I am wondering why SSE pairs are considered service pairs, since they are supported by most services (although look different). 

So, what kind of feature can be considered global?

**Review by xxchan — APPROVED** [review]

**Comment by xxchan** [comment]
What do you mean by 'language related details'? You mean SDKs in different languages? (And so we should follow API, but not SDK)

**Comment by Xuanwo** [comment]
> Why this RFC is numbered 38?

Use the PR number as ID as golang and matrix do.

> BTW, I am wondering why SSE pairs are considered service pairs since they are supported by most services (although look different).

Because different services may have different behavior on it.

For example, Behavior for `Read` global pair `size` and `offset` are the same for every service, but `Write` pair `StorageClass` is different: Depends on service type, the user needs to use different values.

According to SSE supports, the behavior is also very different. S3 supports three types and `SSE-KMS` requires a KMS key id, other services could not support this pair.

In general, If a pair could be used with no service type check, we can consider adding it into global service, otherwise, we need to keep it as service pair.

```go
// size & offset can be used in any service
_, err = store.Read("abc", pairs.WithSize(1024), pairs.Offset(4096))
// But s3 storage class can only be used in s3's store

var pairs []types.Pair
switch tp {
    case s3.Type:
        pairs = append(pairs, s3.WithStorageClass("STANDARD_IA"))
}
_, err = store.Read(“abc”, pairs...)
```

> What do you mean by 'language-related details'? Do you mean SDKs in different languages? (And so we should follow API, but not SDK)

Yes.

> So API Native Style only restricts naming, but not the type/content? Will this cause confusion?

From my point of view, `[]byte` is the real type of this API. Because HTTP headers only allow string, so we have to encode it to base64 like boolean `True` / `False` should be `"true"`/`"false"`. But it doesn't mean we have to set all header-related pairs type to `string`, this makes it much less easy to use.

So we use the `[]byte` and `boolean` for pairs type based on their real type.

**Pull request merged**